### PR TITLE
Add support for the Toggle Snap To Grid module

### PR DESCRIPTION
**Pull request by MAClavell**
I've recently published the [Toggle Snap To Grid](https://github.com/MAClavell/Toggle-Snap-to-Grid) module. To get it working with Drag Ruler I needed to make a couple changes on your end.

## Description
* `options.gridSpaces` is now used in a few places to determine if grid specific things should happen (ex. waypoint centering, measurment coloring and highlighting).
* `dragRulerDeleteWaypoint()` now sets `options.snap` based on `event.shiftKey`. This is so the drag ruler stays gridless when removing waypoints with my grid snapping toggle set to off. I modify `event.shiftKey` in Toggle Snap To Grid to always be true in this case.
* The speed cost API now passes the options parameter passed to Drag Ruler along its entire callstack. Previously, this options object was being created and stored in the `costFunction` in `measureDistances()`, but `getCostFromSpeedProvider()` never passed it along. This was changed to get functionality with Terrain Ruler.
* `terrainRulerAvailable` is now added to the options parameter so it doesn't have to be calculated twice.

## Terrain Ruler Changes
To gain full functionality some code in Terrain Ruler also has to change. The change is small so I didn't want to make a whole fork and PR just for it. The changes are in `measure.js`:

1. Lines 16-19 were changed to: 
```
if (canvas.grid.type === CONST.GRID_TYPES.GRIDLESS || options.ignoreGrid)
    return measureDistancesGridless(segments, options);
else if (canvas.grid.type === CONST.GRID_TYPES.SQUARE)
    return measureDistancesSquare(segments, options);
```
2. Line 168 was changed to:
```
const cost = costFunction((start.x + end.x) / 2, (start.y + end.y) / 2, duplicate(options));
```

### CODE REVIEW

**Comment by manuelVo** [comment]
Thanks for the PR! I've given this a short test and it seems to work well. Before I dive into a deeper code review, there are a few things I'd like to sort out first:

1. Grid highlighting stops working when snapping is being disabled via your module.
2. Instead of overriding `event.shiftKey` in your module to force Drag Ruler into non-snapping mode I'd prefer simply reading whether the toggle from your module is active or not and if it's inactive just set the snap-flag accordingly, ignoring whether shift is pressed or not. (Maybe we could even have shift invert the setting that your module exposes so that if snapping is disabled via your module, shift would temporarily reactivate snapping in Drag Ruler). I see two advantages of doing it that way: 
  i. It would allow me to change the key used to disable snapping in the future without breaking compatibility to your module (I hope to make that key user configurable at some point in the future).
  ii. Your module wouldn't need to modify `event.shiftKey` at all, which decreases the probability of triggering unexpected behavior in other modules.
3. I'm not very fond of switching to euclidean measurement mode when disabling snapping. There are several reasons for this:
  - Switching between snapped mode and non-snapped mode can yield wildly different results for the measurement. For example in 5/5/5 mode, walking two squares diagonally snapped would cost 10ft., while walking the same distance unsnapped would cost ~14ft. (for that cost you could walk almost 3 squares snapped).
  - When disabling snapping *and* using euclidean measurements, I don't see any remaining advantages of using a gridded mode. To accomplish those features, users could simply use gridless to accomplish those features.

What do you think about those points?

**Comment by MAClavell** [comment]
Thanks! I just pushed a few changes as well as an update to the module.

1. Yes grid highlighting is supposed to be disabled when snapping is off. I prefer it this way but in case other people don't I added an option to change if grid highlighting should be active when dragging an unsnapped token.
highlighting-setting
2. Just pushed an update that change how Drag Ruler integration works. Instead of overriding the event when measuring, I take advantage of the 'options' parameter and modify 'snap' and 'gridSpaces' as necessary.
3. As I said above I added an option to switch between grid/highlighted and euclidean/not highlighted when dragging an unsnapped token. I personally prefer grid highlighting off and euclidean distances when moving unsnapped tokens.
    1. Ah I did not know the difference in distance calculation here, but again each user can individually set how grid highlighting works on their clients.
    2. I find it tedious to have to go to the scene settings and switch between gridless and gridded at various points in my games. The goal of my module is to provide players and GMs a way to move tokens how they want.

I just pushed some changes to the branch:
* Pass along the current 'options' parameter in a few places instead of making a new one. This allows outside modules to add and modify options.
* Using the passed in options parameter, check 'options.toggleSnapToGridActive' when adding and removing waypoints.
* Send whether grid snapping is active or not to remote users. This has a couple of benefits:
  * Drag Rulers will now correctly point to where remote players are hovering if they are holding shift during their drag. This feature works regardless of my module being installed or not.
  CorrectUnsnappedPositions
  * With my module modifying snapping and grid highlighting in various ways, this allows the correct ruler state to be shown to all players.

**Comment by manuelVo** [comment]
I'm currently in the process of reviewing the code. That review isn't finished yet, but during the process I've found a few bugs in the current implementation, that I'd like to see fixed before merging.
- When the Terrain Ruler module is disabled, the measurement doesn't switch to Euclidean when it's unsnapped via your module.
- Placing a waypoint while your module forces the token to be unsnapped doesn't place the waypoint at the location where the token is currently hovering, but will place the waypoint in the center of the closest square instead.
- Deleting a waypoint throws an exception.

Additionally, there are two changes that seem to be independent of the "Toggle Snap To Grid Support" feature and would as a result each warrant their own PR (having it split up that way makes it easier to work with the codebase in the future, because it's more apparent why each change in the code was done. What's more is that smaller PRs that each concentrate on a single feature are much easier to review than a single PR with multiple changes in them). Are you able to split those feature up into their separate PRs?

**Comment by MAClavell** [comment]
Pushed some changes and fixes to here and [my module](https://github.com/MAClavell/Toggle-Snap-to-Grid/releases/tag/2.3). I never could get the deleting waypoints exception to occur in my testing, but I changed how that works anyways.

Splitting this PR up is kind of a pain ngl because the changes are mixed together in the commit history. I'll see if I can cherry pick the changes but it most likely will have to be on another day.

Meanwhile here's a list of what I changed:
- When adding and removing a waypoint, check to see if the temporary member variable 'snapOverride' is available. If it is, use that data to determine the snap option instead of the shift key.
    - This is a module agnostic way for anyone to override the snap state. I currently have it implemented in Toggle Snap to Grid.
    - I used a member variable to get around a hoop in Enhanced Terrain Layer where it doesn't recognize the options parameter.
    - This fixes unsnapped waypoints being centered when Terrain Ruler was enabled.
- Use the 'ignoreGrid' option if it is set in the options parameter to determine if Euclidean measurement should be used regardless of what the grid type is.
    - For consistency, this is the same option that Enhanced Terrain Layer uses to force gridless measurement.
    - This lets my module more finely control grid highlighting and measurement type.
    - This fixes Euclidean measurement not working if Terrain Ruler is disabled.

**Comment by manuelVo** [comment]
If splitting the PRs up is too big of a pain in the butt, then leave them together. The changes that could be split away are fairly small, I assume. But let that be a suggestion for the next PR you're making against a repository: The smaller the changeset, the easier and faster it is to review :)

**Comment by manuelVo** [comment]
One remaining bug i found: When having snapping disabled via your module and pressing shift, the mode of measurement changes back go the one that was used for gridded and additionally the highlighting goes back to the gridded-mode as well. Is it somewhat easy to make it so the ruler highlighting and measurement modes stay the same when pressing shift, having the only thing that changes be that the ruler snaps to the grid cell center? If not, simply remove that shift-to-invert option again.

**Comment by MAClavell** [comment]
Haha yeah, I've iterated on these changes so much that it all kind of went out of control.

Not a bug, I purposefully had it revert to the map configuration's measurement and highlight mode when holding shift. Now that you mention it, I can throw that behind a config option as well. This change is purely on my end now that outside modules like mine can control the snapping, measurement, and highlighting state :)

**Comment by manuelVo** [comment]
I'd prefer to not expose any options so other modules can alter it. I think that would introduce a bunch of complexity for very little gain. I'd really like the shift-key to simply change whether the token snaps to the grid or not. That's all it has done in the past and keeping it that way hopefully contributes to user happiness (users really don't tend to like stuff to change). Also, having shift invert snapping behavior temporarily makes some kind of intuitive sense, but I personally see how temporarily switching the measurement scheme for a single drag would make any sense.

File: src/foundry_imports.js [hunk 1]
```diff
@@ -252,8 +271,8 @@ export function measure(destination, {gridSpaces=true, snap=false} = {}) {
 		}
 
 		// Highlight grid positions
-		if (isToken && canvas.grid.type !== CONST.GRID_TYPES.GRIDLESS) {
-			if (terrainRulerAvailable)
+		if (isToken && canvas.grid.type !== CONST.GRID_TYPES.GRIDLESS && options.gridSpaces) {
```
Comment (manuelVo) [inline src/foundry_imports.js:274]: Note to self: I need to think closer about `gridSizes` being used here.

File: src/main.js [hunk 2]
```diff
@@ -171,13 +171,24 @@ function onEntityDragLeftCancel(event) {
 	if (!ruler.isDragRuler || ruler._state === Ruler.STATES.MOVING)
 		return false
 	if (ruler._state === Ruler.STATES.MEASURING) {
+		let options = {};
+
+		// Allow outside modules to override snapping
+		if (this.snapOverride !== undefined && this.snapOverride.active) {
```
Comment (manuelVo) [inline src/main.js:177]: This can be simplified to `if (this.snapOverride?.active)`

File: src/ruler.js [hunk 3]
```diff
@@ -22,12 +22,24 @@ export class DragRulerRuler extends Ruler {
 		// This function is invoked by left clicking
 		if (!this.isDragRuler)
 			return await super.moveToken(event);
+
+		let options = {};
+
+		// Allow outside modules to override snapping
+		if (this.snapOverride !== undefined && this.snapOverride.active) {
```
Comment (manuelVo) [inline src/ruler.js:29]: This is the same code as in `main.js`, isn't it? Let's put that into a function to avoid duplicated code.

File: src/ruler.js [hunk 4]
```diff
@@ -37,6 +49,7 @@ export class DragRulerRuler extends Ruler {
 			const isToken = this.draggedEntity instanceof Token;
 			json["draggedEntityIsToken"] = isToken;
 			json["draggedEntity"] = this.draggedEntity.id;
+			json["snappedToGrid"] = this.snappedToGrid;
```
Comment (manuelVo) [inline src/ruler.js:52]: Is there any reason why other clients would need to know about this? After all, the waypoints that get transmitted are already in their respective snapped or unsnapped position, so what does the other side do with that information? Am I missing something?

File: src/ruler.js [hunk 5]
```diff
@@ -70,8 +88,8 @@ export class DragRulerRuler extends Ruler {
 	}
 
 	// The functions below aren't present in the orignal Ruler class and are added by Drag Ruler
-	dragRulerAddWaypoint(point, snap=true) {
-		if (snap) {
+	dragRulerAddWaypoint(point, options={snap: true}) {
```
Comment (manuelVo) [inline src/ruler.js:91]: This will fail to set `snap: true` as default if options without `snap` in them are passed. Better use an empty object as default and then add `options.snap = options.snap ?? true` to the code below to set that value if it's unset

File: src/ruler.js [hunk 6]
```diff
@@ -91,12 +109,19 @@ export class DragRulerRuler extends Ruler {
 		this.labels.removeChildren().forEach(c => c.destroy());
 	}
 
-	dragRulerDeleteWaypoint(event={preventDefault: () => {return}}) {
+	dragRulerDeleteWaypoint(event={preventDefault: () => {return}}, options={snap: true}) {
```
Comment (manuelVo) [inline src/ruler.js:112]: This will fail to set `snap: true` as default if options without `snap` in them are passed. Better use an empty object as default and then add `options.snap = options.snap ?? true` to the code below to set that value if it's unset

File: src/ruler.js [hunk 7]
```diff
@@ -91,12 +109,19 @@ export class DragRulerRuler extends Ruler {
...
 		if (this.waypoints.filter(w => !w.isPrevious).length > 1) {
 			event.preventDefault();
 			const mousePosition = canvas.app.renderer.plugins.interaction.mouse.getLocalPosition(canvas.tokens);
 			const rulerOffset = this.rulerOffset;
-			this._removeWaypoint({x: mousePosition.x + rulerOffset.x, y: mousePosition.y + rulerOffset.y});
+
+			if(options.snap === undefined){
```
Comment (manuelVo) [inline src/ruler.js:118]: If we already set `snap: true` as default further above, this code shouldn't be necessary.

**Review by manuelVo — CHANGES_REQUESTED** [review]
With most of the features sorted out and working well, here's the first code review.

File: src/ruler.js (hunk 3 above)
Comment (MAClavell) [reply src/ruler.js:29]: Found a typo bug where I was setting 'snapOverrideData' inside of my module instead of 'snapOverride'. Fixed that in v2.4.3

File: src/ruler.js (hunk 4 above)
Comment (MAClavell) [reply src/ruler.js:43]: This change affects how the destination is snapped on remote clients.

**Vanilla Drag Ruler:** This change allows the destination of the ruler to appear unsnapped on remote clients when the local client is holding shift to unsnap. Previously remote clients would see the destination of the ruler snapped to the grid even though that is not what the local client is seeing. This was a nice bonus side effect of the Toggle Snap to Grid integration :)
With my change:
RemotePlayerUnsnap

**With Toggle Snap to Grid installed:** Because of the "hold shift to re-snap" feature my module needs to know if `snap` is currently `false` from holding down shift so it can correctly re-activate it before calling `measure()`. I got this working on local clients but saw that remote clients were not re-snapping because they didn't know if the local client was holding shift or not. This change tells remote clients if the local client is currently snapping to grid and allows my module to apply the correct state.

**Comment by MAClavell** [comment]
> 
> 
> I'd prefer to not expose any options so other modules can alter it. I think that would introduce a bunch of complexity for very little gain. I'd really like the shift-key to simply change whether the token snaps to the grid or not. That's all it has done in the past and keeping it that way hopefully contributes to user happiness (users really don't tend to like stuff to change). Also, having shift invert snapping behavior temporarily makes some kind of intuitive sense, but I personally see how temporarily switching the measurement scheme for a single drag would make any sense.

I personally like it the other way haha. It's behind an option in my module now anyways.

File: src/ruler.js (hunk 4 above)
Comment (manuelVo) [reply src/ruler.js:43]: I just went back and tested this and non-snapping behavior is working without this change - with and without Toggle Snap to Grid enabled. It only seems to break when Toggle Snap to Grid is enabled and the snapping option is enabled for the token. Since all other constellations work, I suppose that that behavior is actually a bug that occurs somewhere in the existing codepath that could be fixed without actually transmitting this additional info.

**Comment by manuelVo** [comment]
Looks pretty good. Seems like there is only the `socketSnappedToGrid`-situation to sort out and then this will be good to go.

Just in case you're wondering about the commit I added to this PR: I noticed that I fucked up on the suggestion for the name of the new `enableTerrainRuler` option and changed the name to what I was actually intending to suggest.

File: src/ruler.js (hunk 4 above)
Comment (MAClavell) [reply src/ruler.js:43]: Just tested it as well and it was partly a bug on my end. My Ruler.measure wrapper was always pushing a default snap state equal to the current grid type. I fixed that, though for compatibility we might still need this change.

Without this change, Drag Ruler sends nothing to remote users about snapping, so options.snap is always undefined when we get to Ruler.measure. That works in our favor because each user is sent the exact coordinate the ruler is pointing to, which the sending client will have already snapped to the grid or left unsnapped if shift is held. Remote clients assume snapping is off at all times and just use whatever coordinate is sent.

Complications appear when my module's logic takes an inputted snap setting and uses that to determine what output to send off to Ruler.measure. A couple of example cases, keep in mind remote clients always assume snapping is disabled:
- Local client has snapping toggled off and is dragging the token normally. My module on the remote client sees `options.snap != true`, assumes shift is being held by the local client, and re-enables snapping. Remote clients see the token snapped to the grid (wrong behavior).
- Local client has snapping toggled off and is holding shift while dragging the token. My module acts the same as above (correct behavior).

Basically, I need some way to know if remote clients are snapping to the grid. Best I can think of is sending that boolean over the network. You have another idea?

Comment (MAClavell) [reply src/ruler.js:43]: Actually just had some ideas. Gonna sleep on it and try some stuff out

Comment (manuelVo) [reply src/ruler.js:43]: In my testing it appears that the state of the "toggle snap to grid" box is already synchronized between the clients. Couldn't you use that value to determine the snapping behavior?

Comment (MAClavell) [reply src/ruler.js]: Pushed an update to my module that removes the need for these changes. Just removed them from this PR as well.

Now it works the same as if my module was uninstalled: a remote client's ruler always has options.snap != true. The destination and all of the waypoints are networked over in the correct state so there isn't a need to know how to snap them. Had to remove that option that allowed reverting to the map's measurement/highlighting when holding shift, but I wasn't ever too attached to it even if that's how I preferred it. Holding shift now just resnaps the token and doesn't modify any more than that.

**Comment by manuelVo** [comment]
Looks good now. Thanks for the PR and thanks for pulling rough the rather long-ish review process with me!

**Comment by manuelVo** [comment]
I'm going to explicitly ask since I've added the LICENSE to the repo after you opened this PR: Are you fine with those changes being published under the MIT license?

**Comment by MAClavell** [comment]
Thanks! Lots of iteration on this one so no problem.

And yeah, MIT license is fine with me.
